include/knowhere/sparse_utils.h

```diff
@@ -0,0 +1,261 @@
+    SparseRow(const SparseRow<T>& other) : SparseRow(other.count_) {
+        std::copy(other.data_, other.data_ + count_ * element_size(), data_);
+    }
+
+    SparseRow(SparseRow<T>&& other) noexcept : SparseRow() {
+        swap(*this, other);
+    }
+
+    SparseRow&
+    operator=(const SparseRow<T>& other) {
+        if (this != &other) {
+            SparseRow<T> tmp(other);
+            swap(*this, tmp);
+        }
+        return *this;
+    }
+
+    SparseRow&
+    operator=(SparseRow<T>&& other) noexcept {
+        swap(*this, other);
+        return *this;
+    }
+
+    ~SparseRow() {
+        if (own_data_ && data_ != nullptr) {
+            delete[] data_;
+            data_ = nullptr;
+        }
+    }
+
+    size_t
+    size() const {
+        return count_;
+    }
+
+    size_t
+    memory_usage() const {
+        return count_ * element_size() + sizeof(*this);
+    }
+
+    void*
+    data() {
+        return data_;
+    }
+
+    const void*
+    data() const {
+        return data_;
+    }
+
+    // dim of a sparse vector is the max index + 1, or 0 for an empty vector.
+    int64_t
+    dim() const {
+        if (count_ == 0) {
+            return 0;
+        }
+        auto* elem = reinterpret_cast<const ElementProxy*>(data_) + count_ - 1;
+        return elem->index + 1;
+    }
+
+    IdVal<T>
+    operator[](size_t i) const {
+        auto* elem = reinterpret_cast<const ElementProxy*>(data_) + i;
+        return {elem->index, elem->value};
+    }
+
+    void
+    set_at(size_t i, table_t index, T value) {
+        auto* elem = reinterpret_cast<ElementProxy*>(data_) + i;
+        elem->index = index;
+        elem->value = value;
+    }
+
+    float
+    dot(const SparseRow<T>& other) const {
+        float product_sum = 0.0f;
+        size_t i = 0, j = 0;
+        while (i < count_ && j < other.count_) {
+            auto* left = reinterpret_cast<const ElementProxy*>(data_) + i;
+            auto* right = reinterpret_cast<const ElementProxy*>(other.data_) + j;
+
+            if (left->index < right->index) {
+                ++i;
+            } else if (left->index > right->index) {
+                ++j;
+            } else {
+                product_sum += left->value * right->value;
+                ++i;
+                ++j;
+            }
+        }
+        return product_sum;
+    }
+
+    friend void
+    swap(SparseRow<T>& left, SparseRow<T>& right) {
+        using std::swap;
+        swap(left.count_, right.count_);
+        swap(left.data_, right.data_);
+        swap(left.own_data_, right.own_data_);
+    }
+
+    static inline size_t
+    element_size() {
+        return sizeof(table_t) + sizeof(T);
+    }
+
+ private:
+    // ElementProxy is used to access elements in the data_ array and should
+    // never be actually constructed.
+    struct __attribute__((packed)) ElementProxy {
+        table_t index;
+        T value;
+        ElementProxy() = delete;
+        ElementProxy(const ElementProxy&) = delete;
```

Comment:
I would add `ElementProxy& operator=(const ElementProxy&) = delete;` as well

Reply:
deleting the default and the copy constructor(so no constructor will be generated) should be sufficient. if no instance can ever be made there is no way to call copy assignment operator.

the reason is this element is used solely to help access what's in the raw `data_` pointer, we don't actually have the need to create such objects.